File: 11-Bridge/src/go/remote.go
package main

import "fmt"

// Abstraction — high-level remote control operations.
type Remote struct {
	device Device // ← THE BRIDGE
}

func NewRemote(device Device) *Remote {
	return &Remote{device: device}
}

func (r *Remote) TogglePower() {
	fmt.Printf("[%s Remote] Toggle power\n", r.device.GetName())
	if r.device.IsEnabled() {
		r.device.Disable()
	} else {
		r.device.Enable()
	}
}

func (r *Remote) VolumeUp() {
	fmt.Printf("[%s Remote] Volume up\n", r.device.GetName())
	r.device.SetVolume(r.device.GetVolume() + 10)
}

func (r *Remote) VolumeDown() {
	fmt.Printf("[%s Remote] Volume down\n", r.device.GetName())
	r.device.SetVolume(r.device.GetVolume() - 10)
}

func (r *Remote) ChannelUp() {
	fmt.Printf("[%s Remote] Channel up\n", r.device.GetName())
	r.device.SetChannel(r.device.GetChannel() + 1)
}

File: 11-Bridge/src/rust/remote.rs
use crate::device::Device;

/// Abstraction — high-level remote control operations.
/// Holds a mutable reference to a Device (the bridge).
pub struct Remote<'a> {
    device: &'a mut dyn Device,
}

impl<'a> Remote<'a> {
    pub fn new(device: &'a mut dyn Device) -> Self {
        Remote { device }
    }

    pub fn toggle_power(&mut self) {
        println!("[{} Remote] Toggle power", self.device.get_name());
        if self.device.is_enabled() {
            self.device.disable();
        } else {
            self.device.enable();
        }
    }

    pub fn volume_up(&mut self) {
        println!("[{} Remote] Volume up", self.device.get_name());
        let vol = self.device.get_volume() + 10;
        self.device.set_volume(vol);
    }

    pub fn volume_down(&mut self) {
        println!("[{} Remote] Volume down", self.device.get_name());
        let vol = self.device.get_volume() - 10;
        self.device.set_volume(vol);
    }

    pub fn channel_up(&mut self) {
        println!("[{} Remote] Channel up", self.device.get_name());
        let ch = self.device.get_channel() + 1;
        self.device.set_channel(ch);
    }
}
